frontend/appflowy_flutter/packages/flowy_infra/lib/plugins/service/models/flowy_dynamic_plugin.dart

```diff
@@ -0,0 +1,137 @@
+  late final String _name;
+
+  String get _fsPluginName => [name, ext].join('.');
+
+  final AppTheme? theme;
+  final String _path;
+
+  Directory get source {
+    return Directory(_path);
+  }
+
+  /// Loads and "compiles" loaded plugins.
+  ///
+  /// If the plugin loaded does not contain the `.flowy_plugin` extension, this
+  /// this method will throw an error. Likewise, if the plugin does not follow
+  /// the expected format, this method will throw an error.
+  static Future<FlowyDynamicPlugin> decode({required Directory src}) async {
+    // throw an error if the plugin does not follow the proper format.
+    if (!isPlugin(src)) {
+      throw PluginCompilationException(
+        'The plugin directory must have the extension `.flowy_plugin`.',
+      );
+    }
+
+    // throws an error if the plugin does not follow the proper format.
+    final type = PluginType.from(src: src);
+
+    switch (type) {
+      case PluginType.theme:
+        return _theme(src: src);
+    }
+  }
+
+  /// Encodes the plugin in memory. The Directory given is not the actual
+  /// directory on the file system, but rather a virtual directory in memory.
+  ///
+  /// Instances of this class should always have a path on disk, otherwise a
+  /// compilation error will be thrown during the construction of this object.
+  Future<Directory> encode() async {
+    final fs = MemoryFileSystem();
+    final result = fs.directory(_fsPluginName)..createSync();
+
+    final lightPath = p.join(_fsPluginName, '$name.$lightExtension');
+    result.childFile(lightPath).createSync();
+    result
+        .childFile(lightPath)
+        .writeAsStringSync(jsonEncode(theme!.lightTheme.toJson()));
+
+    final darkPath = p.join(_fsPluginName, '$name.$darkExtension');
+    result.childFile(darkPath).createSync();
+    result
+        .childFile(p.join(_fsPluginName, '$name.$darkExtension'))
+        .writeAsStringSync(jsonEncode(theme!.darkTheme.toJson()));
+
+    return result;
+  }
+
+  /// Theme plugins should have the following format.
+  /// > directory.flowy_plugin // plugin root
+  /// >   - theme.light.json   // the light theme
+  /// >   - theme.dark.json    // the dark theme
+  ///
+  /// If the theme does not adhere to that format, it is considered an error.
+  static Future<FlowyDynamicPlugin> _theme({required Directory src}) async {
+    late final String name;
+    try {
+      name = p.basenameWithoutExtension(src.path).split('.').first;
+    } catch (e) {
+      throw PluginCompilationException(
+        'The theme plugin does not adhere to the following format: `<plugin_name>.flowy_plugin`.',
+      );
+    }
+
+    final light = src
+        .listSync()
+        .where((event) =>
+            event is File && p.basename(event.path).contains(lightExtension))
+        .first as File;
+
+    final dark = src
+        .listSync()
+        .where((event) =>
+            event is File && p.basename(event.path).contains(darkExtension))
+        .first as File;
```

Comment:
I'm looking at this and I can't help but think if there is a better way to do it.

The only way I could think of introduced ~15 more lines of code though, not sure it's worth it.

<details>
  <summary>~25 lines of code version</summary>

``` 
    final srcAsList = src.listSync();

    File? light;
    File? dark;

    for (int i = 0; i < srcAsList.length; i++) {
      if (light != null && dark != null) {
        break;
      }

      final event = srcAsList[i];
      if (event is File) {
        if (p.basename(event.path).contains(lightExtension)) {
          light = event;
        } else if (p.basename(event.path).contains(darkExtension)) {
          dark = event;
        }
      }
    }

    if (light == null || dark == null) {
      throw PluginCompilationException(
        'The theme plugin does not contain both a light and dark version: `theme.light.json` and `theme.dark.json`.',
      );
    }
``` 

</details>

Reply:
At least this way, we're only listing the directory once. This is a good review comment. Let me try to expand on this.

Reply:
Haha @Xazin, if we're code golfing it then you can try to beat this :)
```dart
    final files = src.listSync().whereType<File>().toList();

    File? light, dark;

    for (int i = 0; i < files.length || (light == null && dark == null); ++i) {
      final file = files[i];
      if (p.basename(file.path).contains(lightExtension)) {
        light = file;
      } else if (p.basename(file.path).contains(darkExtension)) {
        dark = file;
      }
    }
```

Reply:
Either way we still get the benefit of only listing the directory once.

Reply:
Yeah that is indeed better. Semantically it's the same, but more readable.